Add barbican-hsm-plugin interface support

ajkavanagh opened the pull request:
The barbican-hsm-plugin interface provides a mechanism for the Barbican
charm to communicate with an HSM plugin.  The plugin (from the Barbican
perspective) is provided as a PKCS#11 compliant library (.so) and so is
local to the Barbican installation.  Thus, the hsm-plugin charms are
subordinate to the Barbican charm and run on the same unit.

This change also provides two actions (generate-mkek and generate-hmac)
which are 'one-off' operations to initialise the HSM with the global
master keys.

Add a note to the README that the generate-mkek and generate-hmac
actions may only be done once as the HSM may reject overwriting the key.

Add Apache2.0 LICENSE and license headers to files
Removed redundant copyright file

Change the reference for the internal port to 9311 The barbican project
changed the INTERNAL port to the same as the PUBLIC port.

Add in seed_file and seed_length to template. These are needed for a
change in Barbican to support seeding the RNG in the HSM if required.
They are set to /dev/random and 32.

Fetch the barbican sources from a PPA (for bug: 1599550)

Remove the trusty support for Py3 from install hook

ajkavanagh committed:
Add barbican-hsm-plugin interface support

The barbican-hsm-plugin interface provides a mechanism for the Barbican
charm to communicate with an HSM plugin.  The plugin (from the Barbican
perspective) is provided as a PKCS#11 compliant library (.so) and so is
local to the Barbican installation.  Thus, the hsm-plugin charms are
subordinate to the Barbican charm and run on the same unit.

This change also provides two actions (generate-mkek and generate-hmac)
which are 'one-off' operations to initialise the HSM with the global
master keys.

Add a note to the README that the generate-mkek and generate-hmac
actions may only be done once as the HSM may reject overwriting the key.

Add Apache2.0 LICENSE and license headers to files
Removed redundant copyright file

Change the reference for the internal port to 9311 The barbican project
changed the INTERNAL port to the same as the PUBLIC port.

Add in seed_file and seed_length to template. These are needed for a
change in Barbican to support seeding the RNG in the HSM if required.
They are set to /dev/random and 32.

Fetch the barbican sources from a PPA (for bug: 1599550)

Remove the trusty support for Py3 from install hook

javacruft commented on test-requirements.txt:
```diff
@@ -3,4 +3,4 @@ os-testr>=0.4.1
 mock>=1.2
 coverage>=3.6
-git+https://github.com/ajkavanagh/charm.openstack#egg=charms-openstack
+git+https://github.com/ajkavanagh/charms.openstack.git#egg=charms-openstack
```
I think this should be the main openstack version now right?

ajkavanagh replied:
Yes, it should really now.  There's going to be a few of these I think - this'll need a retest on my side.  But, it won't work until we've landed the charms.openstack patchset.

ajkavanagh committed:
Change test-requirements.txt ajkavanagh to openstack

Changed from:
git+https://.../ajkavanagh/charms.openstack.git#egg=charms-openstack
to:
git+https://.../openstack/charms.openstack.git#egg=charms-openstack

This is because the tests should come from the openstack master branch.

ajkavanagh committed:
Add final unit tests

Had missed off unit tests for the mkek and hmac generation code in the
class.

ajkavanagh committed:
Put charm into its own venv

This adds the venv option to the layer configuration and removes the
custom install hook as it's no longer needed.

ajkavanagh committed:
Renamed barbican-hsm-plugin to barbican-hsm

The interface name has been changed, so this ripples it through the charm.
Also fixed the README, and ensured that tox.it has a pep8 target.

gnuoy commented on src/lib/charm/openstack/barbican.py:
```diff
@@ -71,18 +108,59 @@ def __init__(self):
+
 
 class BarbicanAdapters(charms_openstack.adapters.OpenStackRelationAdapters):
     """
```
I think you need OpenStackAPIRelationAdapters here, so that you get the amqp, db and cluster relations.

ajkavanagh replied:
Yes, indeed.  I'll make the change.

gnuoy commented on src/lib/charm/openstack/barbican.py:
```diff
@@ -91,68 +169,141 @@ class BarbicanAdapters(charms_openstack.adapters.OpenStackRelationAdapters):
+        # DEBUG - until seed random change lands into xenial cloud archive
+        # BUG #1599550 - barbican + softhsm2 + libssl1.0.0:
+        #  pkcs11:_generate_random() fails
+        charmhelpers.fetch.add_source("ppa:ajkavanagh/barbican")
```
We can't release the charm into stable with this present. Maybe add a warning here?

ajkavanagh committed:
Change the BarbicanAdapters class parent

charms.openstack has been refactered such that OpenStackRelationAdapters has
a new derived class called OpenStackAPIRelationAdapters which contains the
amqp, db and cluster relations.  As Barbican needs those, the BarbicanAdapters
needs to be derived from the more specialised class.
